--- git status ---
On branch DPPO-Encoder
Your branch is up to date with 'origin/my-fix'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   rsl_rl/runners/on_policy_runner.py

no changes added to commit (use "git add" and/or "git commit -a") 


--- git diff ---
diff --git a/rsl_rl/runners/on_policy_runner.py b/rsl_rl/runners/on_policy_runner.py
index cbd4a3e..e29a450 100644
--- a/rsl_rl/runners/on_policy_runner.py
+++ b/rsl_rl/runners/on_policy_runner.py
@@ -286,12 +286,6 @@ class OnPolicyRunner:
             # Rollout
             with torch.inference_mode():
                 for _ in range(self.num_steps_per_env):
-                    # Implement conventional Network & GRU for adaptation module here
-                    #######################################
-                     # Encode selected observation indices
-
-                    # print(f"Modified observation shape: {modified_obs.shape}")
-                    #######################################
                     # Sample actions
                     # print(modified_obs[0,:])
                     if self.training_type == "dppo":